firehose.go

```diff
@@ -292,10 +309,61 @@ func (fp *Firehose) moveFilesToOutputBucket(ctx context.Context, filePathPrefix
 			// ignore errors when deleting, picked up again on next run
 			_ = fp.StorageBackend.DeleteFile(ctx, fp.processSettings.StagingBucket, fileName)
 		}
+
+		err = fp.writeIndex(ctx, filePathPrefix, currTime, filename, minTimestamp, maxTimestamp)
+		if err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
+// maintain an index file (one for every output directory) listing all the data files and their timestamp range
+func (fp *Firehose) writeIndex(ctx context.Context, filePathPrefix string, currTime time.Time, name string, minTimestamp int64, maxTimestamp int64) error {
+	indexEntry, err := json.Marshal(indexFile{
+		Name:         name,
+		MinTimestamp: minTimestamp,
+		MaxTimestamp: maxTimestamp,
+	})
+	indexEntry = append(indexEntry, '\n')
+	if err != nil {
+		return err
+	}
+	indexLocation := fmt.Sprintf("%s/%s/_metadata.ndjson", filePathPrefix, currTime.Format("2006/1/2"))
+	// GCS objects are immutable: we cannot append to the index. So we read the old index,
+	// append a line, and write a new one. At 5 minutes per entry and a new index file per day,
+	// there are a max of 288 entries per index file, so it isn't too big.
+	index, err := fp.StorageBackend.CreateReader(ctx, fp.processSettings.OutputBucket, indexLocation)
+	var indexContents []byte
+	// If there's an error opening the index, treat it as though it was empty. (Most likely, it didn't exist)
```

Comment:
https://github.com/cloudchacho/hedwig-firehose/blob/main/gcp/gcp.go#L105

Think you can use the precondition to make a reader if it doesn't exist.  Potentially could be another error on createReader and might lose the read data?

Reply:
Thanks for the pointer! I didn't manage to make it work, though: adding `If(storage.Conditions{DoesNotExist: false})` didn't have any effect. (And then I [found](https://pkg.go.dev/cloud.google.com/go/storage#Conditions) that it isn't supposed to do anything.)

I did improve the checking so that other errors won't cause issues.